src/circuit/mont.rs

```diff
@@ -0,0 +1,445 @@
+use pairing::{
+    Engine,
+    Field,
+    PrimeField
+};
+
+use bellman::{
+    SynthesisError,
+    ConstraintSystem,
+    LinearCombination
+};
+
+use super::{
+    Assignment
+};
+
+use super::num::AllocatedNum;
+use super::boolean::{
+    Boolean
+};
+use super::blake2s::blake2s;
+
+use ::jubjub::{
+    JubjubEngine,
+    JubjubParams,
+    montgomery
+};
+
+pub struct MontgomeryPoint<E: Engine, Var> {
+    x: AllocatedNum<E, Var>,
+    y: AllocatedNum<E, Var>
+}
+
+impl<E: JubjubEngine, Var: Copy> MontgomeryPoint<E, Var> {
+    pub fn group_hash<CS>(
+        mut cs: CS,
+        tag: &[Boolean<Var>],
+        params: &E::Params
+    ) -> Result<Self, SynthesisError>
+        where CS: ConstraintSystem<E, Variable=Var>
+    {
+        // This code is specialized for a field of this size
+        assert_eq!(E::Fr::NUM_BITS, 255);
+
+        assert!(tag.len() % 8 == 0);
+
+        // TODO: first block, personalization
+        //
+        // Perform BLAKE2s hash
+        let h = blake2s(cs.namespace(|| "blake2s"), tag)?;
+
+        // Read the x-coordinate
+        let x = AllocatedNum::from_bits_strict(
+            cs.namespace(|| "read x coordinate"),
+            &h[1..]
+        )?;
+
+        // Allocate the y-coordinate given the first bit
+        // of the hash as its parity ("sign bit").
+        let y = AllocatedNum::alloc(
+            cs.namespace(|| "y-coordinate"),
+            || {
+                let s: bool = *h[0].get_value().get()?;
+                let x: E::Fr = *x.get_value().get()?;
+                let p = montgomery::Point::<E, _>::get_for_x(x, s, params);
+                let p = p.get()?;
+                let (_, y) = p.into_xy().expect("can't be the point at infinity");
+                Ok(y)
+            }
+        )?;
+
+        // Unpack the y-coordinate
+        let ybits = y.into_bits_strict(cs.namespace(|| "y-coordinate unpacking"))?;
+
+        // Enforce that the y-coordinate has the right sign
+        Boolean::enforce_equal(
+            cs.namespace(|| "correct sign constraint"),
+            &h[0],
+            &ybits[E::Fr::NUM_BITS as usize - 1]
+        )?;
+
+        // interpret the result as a point on the curve
+        let mut p = Self::interpret(
+            cs.namespace(|| "point interpretation"),
+            &x,
+            &y,
+            params
+        )?;
+
+        // Perform three doublings to move the point into the prime
+        // order subgroup.
+        for i in 0..3 {
+            // Assert the y-coordinate is nonzero (the doubling
+            // doesn't work for y=0).
+            p.y.assert_nonzero(
+                cs.namespace(|| format!("nonzero y-coordinate {}", i))
+            )?;
+
+            p = p.double(
+                cs.namespace(|| format!("doubling {}", i)),
+                params
+            )?;
+        }
```

Comment:
Assert that it's nonzero after the third doubling as well. (I added this to the spec.)

Reply:
AFAIK it's impossible for that to ever take place.

Reply:
You may be right, I'll check.